fixes #3091 - strip leading and trailing whitespace before_save on names of all objects

isratrade opened the pull request:
Replaces PR #884 and strips whitespace on on string attributes, not just :name
This PR also takes care of a feature request to strip whitespace on :mac address.

Commit:
fixes #3091 - strip leading and trailing whitespace before_save on names of all objects

isratrade commented on test/unit/architecture_test.rb:
```diff
@@ -17,10 +17,10 @@ class ArchitectureTest < ActiveSupport::TestCase
+    assert !architecture.name.squeeze(" ").tr(' ', '').empty?
     assert !architecture.save
 
-    architecture.name.strip!.squeeze!(" ").tr!(' ', '')
```
these were removed since they caused failing tests

isratrade commented on app/models/concerns/strip_whitespace.rb:
```diff
@@ -0,0 +1,15 @@
+  extend ActiveSupport::Concern
+
+  included do
+    before_validation :strip_spaces
```
I chose before_validation rather than before_save to prevent validation errors that could error before the callback :strip_spaces is run.  I feel validation should be checked on what is going to be saved to the db.

domcleal commented on app/models/concerns/strip_whitespace.rb:
```diff
@@ -0,0 +1,15 @@
+
+  def strip_spaces
+    self.class.column_names.each do |column|
+      # return string if RuntimeError: can't modify frozen String
```
The return value's not used is it?

domcleal commented:
I've got to admit I'm a little uncomfortable with this being so broad (i.e. affecting every string in the application), but I suspect it is the most effective and consistent way of doing it.

Specifically my concerns are:
a) some fields we might want to exclude, probably config template's "template" field and ptable.  Perhaps others I haven't thought of - maybe automated data like Puppet log messages or reports?
b) we could be changing fields when an object is re-saved, should we just match dirty attributes instead of all columns?

isratrade commented:
@domcleal, good feedback.  I incorporated both suggestions.  I only loop through array of .changed fields and I skip fields in #skip_strip_attrs for each model.

domcleal commented on app/models/concerns/strip_whitespace.rb:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  def strip_spaces
+    self.changed do |column|
```
Nitpick, the `changes` method will provide a hash of the columns, old and new values without needing to use self.send below.

isratrade replied:
@domcleal, I don't think that .strip! on the .changes hash new value is actually going to set the new value for the attribute.

domcleal replied:
ah yes, you might be right.  Though it's in-place modification, so could?

domcleal commented:
I've just analysed my own Foreman database and think the following fields should also be excluded:

Message.value
Ptable.layout

domcleal commented:
@GregSutcliffe, since you probably have a decent sized DB too, would you mind running this on a production (recent nightly) instance and see if you come up with any different answers?

```
ActiveRecord::Base.subclasses.collect { |type| type.name }.sort.reject { |m| m =~ /^(ActiveRecord|Audit)/ }.each { |m| m.constantize.all.each { |r| m.constantize.column_names.each { |c| puts "#{m}.#{c} (#{r.name_method})" if r.send(c).kind_of?(String) && r.send(c).strip != r.send(c) } } }
```

isratrade commented:
@domcleal, new commit that I will squash after your review

GregSutcliffe commented:
@domcleal @isratrade I think you it mostly covered, the only extra one I can see in my db is `KeyPair.secret` which should probably be left verbatim - keys are notoriously picky about format :)

domcleal commented:
Tests too have fallen over again :)

isratrade commented:
@domcleal, I'm confused why the tests failed.  I tried to find the bug but can't debug it now. I will look into it on Sunday.

isratrade commented:
@domcleal, there was a 'typo' in my previous squashed commit where `.each` was removed.  It's back now and the tests pass.  I also added `KeyPair.secret` to be skipped from `.strip!`

domcleal commented:
[test]

isratrade commented:
@domcleal, only failure was 1.9.3 on Pg.  Can you re-run test?
PG::ObjectInUse: ERROR:  database "test_develop_pull_request-2-test" is being accessed by other users
DETAIL:  There are 1 other session(s) using the database.

domcleal commented:
[test] restarted the DB on that slave

skottler commented:
I just tested this out on my development instance which has ~100 hosts (mostly LXC's) checking in and things worked just fine after applying it. The tests are green now, too, so :+1: from me.

domcleal commented:
Merged as e2d3654 for Foreman 1.4, thanks @isratrade.
